### PR TITLE
WebGLPrograms: Improve performance of getParameters().

### PR DESCRIPTION
**Pull request by Mugen87**
I've realized today that cloning uniforms in `getParameters()` should not be necessary. This potential expensive task can be done only when shader program actually needs a compilation. Besides, uniforms are not necessary to compute the program cache key.

@carstenschwede Can you please check if this PR mitigates your performance issue?

### CODE REVIEW

**Comment by Mugen87** [comment]
Too bad, this approach does not work. There is an unfortunate dependency to `Material.onBeforeCompile()`. This callback is executed with the shader object where users have the possibility to modify existing uniforms or add new uniforms. Meaning at the point where `Material.onBeforeCompile()` is called, the final uniforms have to be ready. But since `Material.onBeforeCompile()` can alter the shader code, we can't call it after `getParamters()`.

**Comment by carstenschwede** [comment]
Thanks for the effort! Unfortunately I still see light state changes in a scene rendering only a `CubeCamera`.

**Comment by Mugen87** [comment]
> Thanks for the effort! Unfortunately I still see light state changes in a scene rendering only a CubeCamera.

Yes, but the overhead related to `getUniforms()` should be gone.

**Comment by carstenschwede** [comment]
Sorry, you are right of course:

Before:
Screen Shot 2020-06-16 at 16 53 06

With this PR:
Screen Shot 2020-06-16 at 16 56 58

**Comment by Mugen87** [comment]
With this PR, code like this does not work anymore:

https://github.com/mrdoob/three.js/blob/681ba71ef0c5ba8c383b3e795fd62477929217b7/examples/webgl_materials_modified.html#L42

However, I'm doing some further testing and maybe there is a way to speed up the implementation without introducing breakage.

**Comment by Mugen87** [comment]
Sorry, but I see no way to fix this without refactoring `NodeMaterial`. Even with the previous commit, node examples like `webgl_sprites_nodes` break. See #14442 and related issues/PRs for more information .

@mrdoob In the meanwhile, it seems to me `NodeMaterial` somewhat misused `Material.onBeforeCompile()`for its build system. Hence, it was necessary to refactor the usage of `Material.onBeforeCompile()` in the past so that node material works. 

And now, even the name `onBeforeCompile()` is wrong since the callback is executed not before each shader compilation but in each invocation of `getParameters()`. 

This will definitely be a problem if the engine wants to have multiple programs per material, see #15047.

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -1536,6 +1536,10 @@ function WebGLRenderer( parameters ) {
 
 		if ( programChange ) {
 
+			parameters.uniforms = programCache.getUniforms( material, parameters );
+
+			material.onBeforeCompile( parameters, _this );
```
Comment (Mugen87) [inline src/renderers/WebGLRenderer.js:1541]: Need to pass in `parameters` so user code actually modifies the correct reference to `vertexShader` and `fragmentShader`.

**Comment by Mugen87** [comment]
While we are at it: #17567 should be merged to solve a conceptual issue with `onBeforeCompile()`.

If  the user implements `onBeforeCompile()` with dynamic code (e.g. the output depends on if/else or switch statements), the following statement fails since `toString()` always returns the same result:

https://github.com/mrdoob/three.js/blob/681ba71ef0c5ba8c383b3e795fd62477929217b7/src/renderers/webgl/WebGLPrograms.js#L339

Merging #17567 might provide an API for `NodeMaterial` in order to distinct shader programs.

**Comment by mrdoob** [comment]
> Merging #17567 might provide an API for NodeMaterial in order to distinct shader programs.

Merged! 👍

**Comment by mrdoob** [comment]
I've not studied this PR yet but, just so you know, [I've updated](https://github.com/mrdoob/three.js/commit/3ecbfee3e4c3ac87457eef4b9147d1366f9becc8) `webgl_materials_modified` to reflect #17567:

**Comment by Mugen87** [comment]
>  I've not studied this PR yet but, just so you know, I've updated webgl_materials_modified to reflect #17567:

Cool! 👍 

This whole topic is quite complicated and I had to invest a lot of time today in order to recall all the details (except from `NodeMaterial`'s implementation)^^. It might be necessary to target `r119` if `NodeMaterial` shouldn't break with this change.

**Comment by Mugen87** [comment]
@sunag I think `NodeMaterial` needs to be refactored so that the call of `NodeMaterial.build()` does not happen in `onBeforeCompile()`.

https://github.com/mrdoob/three.js/blob/5dbfa90a4e606a0bab31989fa064a5235dd4ffaa/examples/jsm/nodes/materials/NodeMaterial.js#L31-L43

Instead, `NodeMaterial` should provide a method that tells the renderer whether to trigger a build or not in `initMaterial()` right before `getParameters()` is called. Something like:
```js
if ( material.isNodeMaterial && material.requiresBuild() === true ) {

    material.build( { renderer: _this } );

}

const parameters = programCache.getParameters( material, lights.state, shadowsArray, scene, _clipping.numPlanes, _clipping.numIntersection, object );
const programCacheKey = programCache.getProgramCacheKey( parameters );
```
`NodeMaterial`'s implementation of `onBeforeCompile()` can then be reduced to just this:
```js
this.onBeforeCompile = function ( shader ) {

	shader.uniforms = this.uniforms;
	shader.vertexShader = this.vertexShader;
	shader.fragmentShader = this.fragmentShader;

};
```
The implementation of `requiresBuild()` could produce some sort of internal version counter based on the current node configuration.

**Comment by Mugen87** [comment]
@sunag It seems `NodeMaterial` needs the ability to quickly compute a hash that represents the current node configuration. This hash value could be returned in the new callback `Material.customProgramCacheKey()` in order to identify the shader program for a node material.

Meaning instead of having this code in `NodeMaterial`:

https://github.com/mrdoob/three.js/blob/5dbfa90a4e606a0bab31989fa064a5235dd4ffaa/examples/jsm/nodes/materials/NodeMaterial.js#L47-L53

it should be:

```js
this.customProgramCacheKey = function () {

    return this.computeHash();
 
};
```

**Comment by carstenschwede** [comment]
Is there currently an example that fails using this PR? I had some issues using `GLTFLoader()` but those seem to be fixed in the current state of this PR.

**Comment by Mugen87** [comment]
> Is there currently an example that fails using this PR?

Yes, it's `webgl_sprites_nodes`. This example exposes a problem in `NodeMaterial` described above.

**Comment by Mugen87** [comment]
Related `NodeMaterial` issue: #18347

**Comment by sunag** [comment]
@Mugen87 Interesting... `Material.customProgramCacheKey` seems very good. This PR break only `NodeMaterial` example?

**Comment by Mugen87** [comment]
Yes. But it seems only `webgl_sprites_nodes` is affected. This example actually exposes the problem that the renderer can't identify the correct program for `NodeMaterial`.

**Comment by sunag** [comment]
It seems a good opportunity to create a `Node.getHash()`.

**Comment by Mugen87** [comment]
Sounds good! This should also fix #18347.

**Comment by al3xpisani** [comment]
hi, what needs to be done to fix this issue pls ? Im facing this error on my project

TypeError: d.customProgramCacheKey is not a functionthree.min.js:103:324
    getParameters https://threejs.org/build/three.min.js:103
    p https://threejs.org/build/three.min.js:199
    k https://threejs.org/build/three.min.js:204
    renderBufferDirect https://threejs.org/build/three.min.js:219
    onBeforeRender http://192.168.15.3:8080/js/Lensflare.js:226
    q https://threejs.org/build/three.min.js:198
    m https://threejs.org/build/three.min.js:198
    render https://threejs.org/build/three.min.js:225
    render http://192.168.15.3:8080/js/app.js:630
    init http://192.168.15.3:8080/js/app.js:172
    Kd https://threejs.org/build/three.min.js:222
    a https://threejs.org/build/three.min.js:28

**Comment by Mugen87** [comment]
You need to share some more details about this issue. Do you mind demonstrating it with a live example? https://jsfiddle.net/hyok6tvj/

**Comment by al3xpisani** [comment]
> You need to share some more details about this issue. Do you mind demonstrating it with a live example? https://jsfiddle.net/hyok6tvj/

sure. [https://codesandbox.io/s/elated-bas-jm9ze]

its on /js/Light/Lights.js 

If I comment out line 93 (    scene.add(light2);  ) it works .

**Comment by Mugen87** [comment]
Sorry, but I'm afraid your app setup is incorrect. You include `three.js` (via `https://threejs.org/build/three.js`) _AND_ `js/three.module.js` which is invalid. You should always have just a single version of `three.js` in your build. 

Another thing which is unrelated to your issue: You use the latest `three.js` version but then include `OrbitControls` from `r85`(via https://unpkg.com/three@0.85.0/examples/js/controls/OrbitControls.js). This is also not valid. All library files have to be from the same release.

Please use the [forum](https://discourse.threejs.org/) if you need help to clean up these inconsistencies.

**Comment by al3xpisani** [comment]
> Sorry, but I'm afraid your app setup is incorrect. You include `three.js` (via `https://threejs.org/build/three.js`) _AND_ `js/three.module.js` which is invalid. You should always have just a **single** version of `three.js` in your build. Besides, it's best to focus on a ES6 modules build and avoid the usage of global scripts if possible.
> 
> Another thing which is unrelated to your issue: You use the latest `three.js` version but then include `OrbitControls` from `r85`(via https://unpkg.com/three@0.85.0/examples/js/controls/OrbitControls.js). This is also not valid. All library files have to be from the same release.
> 
> Please use the [forum](https://discourse.threejs.org/) if you need help to clean up these inconsistencies.

Sure. I see and thank you for helping me. Im just starting at this threejs and js world. will get updates :)

**Comment by mrdoob** [comment]
Thanks!

**Comment by Mugen87** [comment]
Yay! 🎉

**Pull request merged**